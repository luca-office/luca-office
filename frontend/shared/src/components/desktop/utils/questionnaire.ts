import {ScenarioQuestionnairesQuery_scenarioQuestionnaires} from "../../../graphql/generated/ScenarioQuestionnairesQuery"

export const questionnaireTypeForQuestionId = (
  questionId: UUID,
  questionnaireId: UUID,
  scenarioQuestionnaires: ScenarioQuestionnairesQuery_scenarioQuestionnaires[]
) =>
  scenarioQuestionnaires
    .filter(scenarioQuestionnaire => scenarioQuestionnaire.questionnaireId === questionnaireId)
    .map(scenarioQuestionnaires => scenarioQuestionnaires.questionnaire)
    .flatMap(questionnaire => questionnaire.questions)
    .find(question => question.id === questionId)?.questionType
