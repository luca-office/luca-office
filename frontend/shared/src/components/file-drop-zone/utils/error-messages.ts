import {LucaI18nLangKey} from "../../../translations"

export enum FileUploadError {
  FileTooLarge,
  InvalidFileType,
  TooManyFiles
}

export const getLanguageKeyForFileUploadError = (error: FileUploadError): LucaI18nLangKey => {
  switch (error) {
    case FileUploadError.FileTooLarge:
      return "files_and_directories__upload_modal_file_too_large"
    case FileUploadError.InvalidFileType:
      return "files_and_directories__upload_modal_file_type_invalid"
    case FileUploadError.TooManyFiles:
      return "files_and_directories__upload_modal_no_single_file"
  }
}

export type FileErrorMessage = "file-too-large" | "file-too-small" | "too-many-files" | "file-invalid-type"

export const mapDropzoneRejectionToFileUploadError = (errorMessage: FileErrorMessage) => {
  switch (errorMessage) {
    case "file-invalid-type":
      return FileUploadError.InvalidFileType
    case "file-too-large":
      return FileUploadError.FileTooLarge
    case "too-many-files":
      return FileUploadError.TooManyFiles
    default:
      return FileUploadError.InvalidFileType
  }
}
