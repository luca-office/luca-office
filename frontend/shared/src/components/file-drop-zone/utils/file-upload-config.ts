export const maximumFileSizeInMegaByte = 100
export const maximumFileSizeInBytes = maximumFileSizeInMegaByte * Math.pow(10, 6)
