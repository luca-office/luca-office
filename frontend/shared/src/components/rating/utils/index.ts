export * from "./coding-criterion"
export * from "./coding-dimension"
export * from "./coding-item"
export * from "./question"
export * from "./score"
export * from "./scoring-type"
