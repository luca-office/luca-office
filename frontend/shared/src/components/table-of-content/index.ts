export * from "./table-of-contents"
export * from "./table-of-contents-entry/table-of-contents-entry"
