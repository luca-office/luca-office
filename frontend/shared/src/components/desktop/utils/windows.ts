import findLastIndex from "lodash-es/findLastIndex"
import {OfficeTool, OfficeWindowType} from "../../../enums"
import {Option} from "../../../utils/option"

export const findVisibleWindows = (openedWindows: OfficeWindowType[], minimizedWindows: OfficeWindowType[]) => {
  const visibleWindows = openedWindows.filter(window => !minimizedWindows.includes(window))

  return Option.of(findLastIndex(visibleWindows, isFullScreenWindow))
    .map(lastFullScreenToolIndex => visibleWindows.slice(lastFullScreenToolIndex))
    .getOrElse(visibleWindows)
}

const isFullScreenWindow = (window: OfficeWindowType) => {
  switch (window) {
    case OfficeTool.Notes:
    case OfficeTool.Calculator:
    case OfficeTool.Chat:
      return false
    default:
      return true
  }
}
