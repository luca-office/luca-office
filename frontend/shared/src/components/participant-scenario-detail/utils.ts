import {FreetextAnswerFragment} from "../../graphql/generated/FreetextAnswerFragment"
import {QuestionType} from "../../graphql/generated/globalTypes"
import {Questionnaire, RuntimeSurveyResults} from "../../models"
import {ParticipantQuestionAnswer, QuestionnaireParticipantResults} from "../questionnaire"

export const runtimeSurveyResultsToQuestionnaireParticipantResults = (
  invitationId: UUID,
  questionnaire: Questionnaire,
  runtimeSurveyResults: RuntimeSurveyResults,
  freeTextAnswers: FreetextAnswerFragment[]
): QuestionnaireParticipantResults => {
  const answers = runtimeSurveyResults.questionResults.reduce<Record<UUID, ParticipantQuestionAnswer>>(
    (answersAccumulator, result) => {
      const question = questionnaire.questions.find(question => question.id === result.questionId)
      const participantResult = result.participantResults.find(results => results.invitationId === invitationId)
      const freeTextAnswer = freeTextAnswers.find(
        answer => answer.questionId === result.questionId && answer.surveyInvitationId === invitationId
      )?.text

      switch (question?.questionType) {
        case QuestionType.SingleChoice: {
          return {
            ...answersAccumulator,
            [question.id]: {
              questionType: QuestionType.SingleChoice,
              questionId: question.id,
              selectedAnswerId: participantResult?.selectedAnswerIds[0] ?? "",
              isFreetextAnswerSelected: participantResult?.wasFreetextAnswerSelected ?? false,
              freeTextAnswer: freeTextAnswer
            }
          }
        }

        case QuestionType.MultipleChoice: {
          return {
            ...answersAccumulator,
            [question.id]: {
              questionType: QuestionType.MultipleChoice,
              questionId: question.id,
              selectedAnswerIds: participantResult?.selectedAnswerIds ?? [],
              isFreetextAnswerSelected: participantResult?.wasFreetextAnswerSelected ?? false,
              freeTextAnswer: freeTextAnswer
            }
          }
        }

        case QuestionType.FreeText: {
          return {
            ...answersAccumulator,
            [question.id]: {
              questionType: QuestionType.FreeText,
              questionId: question.id,
              freeTextAnswer: freeTextAnswer ?? ""
            }
          }
        }

        default:
          return answersAccumulator
      }
    },
    {}
  )

  return {questionnaireId: runtimeSurveyResults.questionnaireId, answers}
}
