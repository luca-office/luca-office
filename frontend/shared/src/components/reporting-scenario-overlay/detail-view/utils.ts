import {TableEntity} from "../../../components/rating/models"
import {CodingDimension, CodingItem, CodingItemResultByItemId} from "../../../models"

export const itemToTableEntity = (
  item: CodingItem,
  codingItemResultsByItem: CodingItemResultByItemId
): TableEntity => ({
  id: item.id,
  title: item.title,
  position: item.position,
  score: codingItemResultsByItem[item.id].score,
  maxScore: codingItemResultsByItem[item.id].maximumScore,
  averageScore: codingItemResultsByItem[item.id].averageScore,
  rated: true
})

export const dimensionToTableEntity = (
  dimension: CodingDimension,
  items: CodingItem[],
  codingItemResultsByItem: CodingItemResultByItemId
): TableEntity => ({
  id: dimension.id,
  title: dimension.title,
  position: dimension.position,
  score: items.reduce((acc, item) => acc + codingItemResultsByItem[item.id].score, 0),
  maxScore: items.reduce((acc, item) => acc + codingItemResultsByItem[item.id].maximumScore, 0),
  averageScore: items.reduce((acc, item) => acc + codingItemResultsByItem[item.id].averageScore, 0),
  rated: true
})
