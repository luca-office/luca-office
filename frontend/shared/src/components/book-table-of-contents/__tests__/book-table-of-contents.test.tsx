import {shallow} from "enzyme"
import * as React from "react"
import {create} from "react-test-renderer"
import {fakeTranslate} from "../../../../tests/utils/translate-mock"
import {Option} from "../../../utils"
import {BookTableOfContents, BookTableOfContentsProps, Button, Card, CardFooter, CardHeader} from "../.."
import {TableOfContentsContainer, TableOfContentsEntry} from "../../table-of-content"
import {tocBookMock} from "../__mocks__/toc-book-data.mock"

const selectSpy = jest.fn()

const defaultProps: BookTableOfContentsProps = {
  title: "Title",
  bookChapters: tocBookMock,
  selectEntityId: selectSpy,
  selectedEntityId: Option.none(),
  t: fakeTranslate
}

const getComponent = (props?: Partial<React.PropsWithChildren<BookTableOfContentsProps>>) => (
  <BookTableOfContents {...{...defaultProps, ...props}} />
)

describe("book-table-of-contents", () => {
  it("renders correctly", () => {
    const component = getComponent()
    const tree = create(component).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("has correct structure", () => {
    const component = shallow(getComponent())
    const tocContainer = component.find(TableOfContentsContainer)
    expect(tocContainer).toHaveLength(1)
    expect(tocContainer.find(TableOfContentsEntry)).toHaveLength(2)

    expect(tocContainer.dive().find(CardHeader)).toHaveLength(1)
    expect(tocContainer.dive().find(CardFooter)).toHaveLength(1)
  })

  it("has correct structure with extra elements", () => {
    const addSpy = jest.fn()
    const component = shallow(
      getComponent({
        addButtonConfig: {labelKey: "placeholder", onClick: addSpy},
        actionFooter: <div className={"test-div"}>foo</div>
      })
    )

    const tocContainer = component.find(TableOfContentsContainer)
    expect(tocContainer).toHaveLength(1)
    expect(tocContainer.find(TableOfContentsEntry)).toHaveLength(2)

    expect(tocContainer.dive().find(Card)).toHaveLength(1)
    expect(tocContainer.dive().find(CardHeader)).toHaveLength(1)
    expect(tocContainer.dive().find(CardFooter)).toHaveLength(1)
    expect(tocContainer.dive().find(".test-div")).toHaveLength(1)
    expect(tocContainer.dive().find(Button)).toHaveLength(1)
    tocContainer.dive().find(Button).simulate("click")
    expect(addSpy).toHaveBeenCalled()
  })
  it("has correct structure with hidden chapters", () => {
    const component = shallow(getComponent({hideChapters: true}))
    const tocContainer = component.find(TableOfContentsContainer)
    expect(tocContainer).toHaveLength(1)
    expect(tocContainer.find(TableOfContentsEntry)).toHaveLength(2)

    expect(tocContainer.dive().find(CardHeader)).toHaveLength(1)
    expect(tocContainer.dive().find(CardFooter)).toHaveLength(1)
  })
})
