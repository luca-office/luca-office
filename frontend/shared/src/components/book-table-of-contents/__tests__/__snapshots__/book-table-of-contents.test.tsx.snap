// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`book-table-of-contents renders correctly 1`] = `
<div
  className="css-1tz1azl-Card"
>
  <div
    className="card-header css-15zaw2t-CardHeader"
  >
    <div
      className="css-mv0otu-TableOfContentsContainer"
    >
      <div
        className="css-1f6p2vj-Heading"
      >
        Title
      </div>
    </div>
  </div>
  <div
    className="toc-body css-awggpr-TableOfContentsContainer"
  >
    <div
      className="book-chapters-list css-1n23nny-TableOfContentsContainer"
    >
      <div
        className="css-18xeog2-TableOfContentsContainer"
      >
        <div
          className="toc-entry css-12opz0x-TableOfContentsEntry"
        >
          <div
            className="css-1ldtu9f-Icon"
            dangerouslySetInnerHTML={
              {
                "__html": {},
              }
            }
            onClick={[Function]}
          />
          <div
            className="toc-entry-text css-x4fs04-TableOfContentsEntry"
            onClick={[Function]}
          >
            <div
              className="css-scbjzw-Icon"
              dangerouslySetInnerHTML={
                {
                  "__html": {},
                }
              }
            />
            <div
              className="css-1k0t96a-Text"
            >
              animal treatments
            </div>
          </div>
        </div>
        <div
          className="toc-entry css-12opz0x-TableOfContentsEntry"
        >
          <div
            className="css-1ldtu9f-Icon"
            dangerouslySetInnerHTML={
              {
                "__html": {},
              }
            }
            onClick={[Function]}
          />
          <div
            className="toc-entry-text css-x4fs04-TableOfContentsEntry"
            onClick={[Function]}
          >
            <div
              className="css-scbjzw-Icon"
              dangerouslySetInnerHTML={
                {
                  "__html": {},
                }
              }
            />
            <div
              className="css-1k0t96a-Text"
            >
              dinosaur breeds
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="card-footer css-105jaxv-CardFooter"
  />
</div>
`;
